lib/sparkql/evaluator.rb

```diff
@@ -0,0 +1,152 @@
+# Using an external source for resolving the individual sparkql expressions,
+# this class will evaluate the rest of a parsed sparkql string to true or false.
+# Namely, this class will handle all the nesting, boolean algebra, and dropped 
+# fields. Plus, it has some optimizations built in to skip the processing for
+# any expressions that don't contribute to the net result of the filter.
+class Sparkql::Evaluator
+
+  attr_reader :processed_count
+
+  def initialize expression_resolver
+    @resolver = expression_resolver
+  end
+
+  def evaluate(expressions)
+    @processed_count = 0
+    @index = {
+      level: 0,
+      block_group: 0,
+      conjunction: "And",
+      conjunction_level: 0,
+      match: true,
+      good_ors: false,
+      expressions: 0
+    }
+    @groups = [@index]
+    expressions.each do |expression|
+      handle_group(expression)
+      adjust_expression_for_dropped_field(expression)
+      check_for_good_ors(expression)
+      next if skip?(expression)
+      result = evaluate_expression(expression)
+    end
+    cleanup
+    return @index[:match]
```

Comment:
Since I believe this is the same as what cleanup returns, and `cleanup` feels kind of like a misnomer, maybe just collapse this to a single line and rename the method to note that it evaluates the final result?